test-data/unit/check-unreachable-code.test

```diff
@@ -457,3 +458,44 @@ else:
 reveal_type(x)  # E: Revealed type is 'builtins.str'
 [builtins fixtures/ops.pyi]
 [out]
+
+[case testShortCircuitInExpression]
+import typing
+def f() -> bool: pass
+PY2 = f()
+PY3 = f()
+
+a = PY2 and 's'
+b = PY3 and 's'
+c = PY2 or 's'
+d = PY3 or 's'
+reveal_type(a)  # E: Revealed type is 'builtins.bool'
+reveal_type(b)  # E: Revealed type is 'builtins.str'
+reveal_type(c)  # E: Revealed type is 'builtins.str'
+reveal_type(d)  # E: Revealed type is 'builtins.bool'
+[builtins fixtures/ops.pyi]
+[out]
+
+[case testShortCircuitAndWithConditionalAssignment]
+# flags: --platform linux
+import sys
+
+def f(): pass
+PY2 = f()
+if PY2 and sys.platform == 'linux':
+    x = 'foo'
+else:
+    x = 3
+reveal_type(x)  # E: Revealed type is 'builtins.int'
+
+[case testShortCircuitOrWithConditionalAssignment]
+# flags: --platform linux
+import sys
+
+def f(): pass
+PY2 = f()
+if PY2 or sys.platform == 'linux':
+    x = 'foo'
+else:
+    x = 3
+reveal_type(x)  # E: Revealed type is 'builtins.str'
```

Comment:
I would also add few more tests here and for ``and``, so that more combinations of ``True`` and ``False`` are tested.